Add COVAR_POP, COVAR_SAMP, CORR, REGR_SLOPE, REGR_INTERCEPT

haozhun opened the pull request:
Supersedes #2051

haozhun commented:
@electrum This pull request should have fixed all comments you pointed out in #2051. Otherwise, it is the same as #2051 except for bring everything up to date.

electrum commented:
`covar_pop` seems to be missing

electrum commented:
Add release notes

electrum commented on presto-docs/src/main/sphinx/functions/aggregate.rst:
```diff
@@ -139,6 +139,22 @@ Approximate Aggregate Functions
 Statistical Aggregate Functions
 -------------------------------
 
+.. function:: corr(x, y) -> double
```
This is missing `covar_pop`

haozhun replied:
It's there now

cberner replied:
All these arguments are in the wrong order. `y` should be the first argument. See `binary set function` in the grammer of the SQL spec

haozhun commented:
* Fixed mathematical bug in original code
* Made tests non-trivial
* Added missing covar_pop
* Made edge case handling agree to standard
* Simplified math

haozhun commented:
Ready for review

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/CorrelationAggregation.java:
```diff
@@ -0,0 +1,61 @@
+    @InputFunction
+    public static void input(CorrelationState state, @SqlType(StandardTypes.DOUBLE) double independentValue, @SqlType(StandardTypes.DOUBLE) double dependentValue)
+    {
+        updateCorrelationState(state, independentValue, dependentValue);
```
name these `x` and `y`

cberner replied:
nevermind, the spec uses those terms

haozhun replied:
OK. That's where independent and dependent come from.

cberner replied:
Swap the names. The standard refers to the first value as the dependent value

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/CovarianceAggregation.java:
```diff
@@ -0,0 +1,67 @@
+    private CovarianceAggregation() {}
+
+    @InputFunction
+    public static void input(CovarianceState state, @SqlType(StandardTypes.DOUBLE) double independentValue, @SqlType(StandardTypes.DOUBLE) double dependentValue)
```
Swap the order of the arguments

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/RegressionAggregation.java:
```diff
@@ -0,0 +1,78 @@
+    @InputFunction
+    public static void input(RegressionState state, @SqlType(StandardTypes.DOUBLE) double independentValue, @SqlType(StandardTypes.DOUBLE) double dependentValue)
+    {
+        updateRegressionState(state, independentValue, dependentValue);
```
This is incorrect. The first argument is the dependent value

cberner commented on presto-docs/src/main/sphinx/functions/aggregate.rst:
```diff
@@ -139,6 +139,26 @@ Approximate Aggregate Functions
+
+.. function:: regr_intercept(x, y) -> double
+
+    Returns linear regression intercept of input values.
```
clarify that y is the dependent variable and swap the order of the arguments

cberner commented on presto-docs/src/main/sphinx/functions/aggregate.rst:
```diff
@@ -139,6 +139,26 @@ Approximate Aggregate Functions
+
+.. function:: regr_slope(x, y) -> double
+
+    Returns linear regression slope of input values.
```
clarify that y is the dependent variable and swap the order of the arguments

cberner commented:
Assign back to me, after you've fixed the argument order

cberner commented:
Couple minor comments. Looks good
